cardano-db-sync/src/Cardano/DbSync/Util/Constraint.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.DbSync.Util.Constraint (
+  constraintNameEpochStake,
+  constraintNameReward,
+  dbConstraintNamesExists,
+  addConstraintsIfNotExist,
+  addRewardTableConstraint,
+  addEpochStakeTableConstraint,
+) where
+
+import Cardano.Db (ManualDbConstraints (..))
+import qualified Cardano.Db as DB
+import Cardano.DbSync.Api.Types (SyncEnv (..))
+import Cardano.Prelude (MonadIO (..), Proxy (..), ReaderT (runReaderT), atomically, when)
+import Control.Concurrent.Class.MonadSTM.Strict (readTVarIO, writeTVar)
+import Control.Monad (unless)
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Database.Persist.Names (ConstraintNameDB (..), FieldNameDB (..))
+import Database.Persist.Postgresql (PersistEntity (..), SqlBackend)
+
+constraintNameEpochStake :: ConstraintNameDB
+constraintNameEpochStake = ConstraintNameDB "unique_epoch_stake"
+
+constraintNameReward :: ConstraintNameDB
+constraintNameReward = ConstraintNameDB "unique_reward"
+
+-- We manually create unique constraints to improve insert speeds when syncing
+-- This function checks if those constraints have already been created
+dbConstraintNamesExists :: MonadIO m => SqlBackend -> m ManualDbConstraints
+dbConstraintNamesExists sqlBackend = do
+  runReaderT queryRewardAndEpochStakeConstraints sqlBackend
+
+queryRewardAndEpochStakeConstraints ::
+  MonadIO m =>
+  ReaderT SqlBackend m ManualDbConstraints
+queryRewardAndEpochStakeConstraints = do
+  resEpochStake <- DB.queryHasConstraint constraintNameEpochStake
+  resReward <- DB.queryHasConstraint constraintNameReward
+  pure $
+    ManualDbConstraints
+      { dbConstraintRewards = resReward
+      , dbConstraintEpochStake = resEpochStake
+      }
+
+addConstraintsIfNotExist ::
+  forall m.
+  (MonadBaseControl IO m, MonadIO m) =>
+  SyncEnv ->
+  ReaderT SqlBackend m ()
+addConstraintsIfNotExist syncEnv = do
+  ManualDbConstraints {..} <- liftIO . readTVarIO $ envDbConstraints syncEnv
+  unless dbConstraintRewards addRewardTableConstraint
+  unless dbConstraintEpochStake addEpochStakeTableConstraint
+  when (dbConstraintRewards && dbConstraintEpochStake) $
```

Comment:
This should always update. The constraints are always created at this point, so the in memory flags should match this.